libs/sync_irq/src/lib.rs

```diff
@@ -0,0 +1,24 @@
+#![feature(negative_impls)]
+#![no_std]
+
+pub type Mutex<T> = sync::Mutex<DisableIrq, T>;
+pub type MutexGuard<'a, T> = sync::MutexGuard<'a, DisableIrq, T>;
+
+/// A deadlock prevention method that disables interrupt requests.
+#[derive(Copy, Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct DisableIrq {}
+
+impl !Send for DisableIrq {}
+
+impl sync::DeadlockPrevention for DisableIrq {
+    #[inline]
+    fn enter() {
+        // FIXME: Recursive disabling doesn't work.
```

Comment:
yeah, i think this is a symptom of this trait design, where `DeadlockPrevention`'s methods are a bit too simple. See my comment there.